### ISSUE TITLE
charge battery from the grid

### ISSUE THREAD

**Issue by kicksev** [biemond/solaredge.modbus#57]
Hi, I have a Solaredge 3f 10,5K system with a battery and I am using the SolarEdge + Growatt TCP modbus app in my Homey Pro. I am looking for a way to charge the battery from the grid. When setting Storage Control to "remote control"and Storage charge/discharge mode to "charge from PV+AC", the house is charged from the grid and the battery is only charged from the PV. As the PV power becomes 0, the house is still charged from the grid and the battery is discharging to the house with a tiny 0,05kW. The battery never gets charged from the grid. What modes should I set to make the battery being charged from the grid?

**Comment by biemond** [comment]
works for me 

<img width="433" alt="Screen Shot 2023-10-22 at 19 20 21" src="https://github.com/biemond/solaredge.modbus/assets/1085004/d29e2d6e-aa20-4d96-bcd7-e6561e1fc27d">
<img width="408" alt="Screen Shot 2023-10-22 at 19 20 48" src="https://github.com/biemond/solaredge.modbus/assets/1085004/8851a49f-5958-4c0a-8e08-25887fe96504">
<img width="398" alt="Screen Shot 2023-10-22 at 19 33 55" src="https://github.com/biemond/solaredge.modbus/assets/1085004/8a2ce41d-13ea-4fc5-aa66-fe94e0ea750f">

**Comment by biemond** [comment]
maybe ask for a firmware update

**Comment by CaptainVoni** [comment]
I have the same situation here. Grid charging is simply not enabled in some regions it seems. I don't even have the option to enable grid charging in the SetApp.

@biemond What inverter and what firmware do you have exactly?

**Comment by kicksev** [comment]
A few days ago out of the blue all functions started to run. I assume there is new software loaded into my inverter from remote or some magic happened. With Storage Contol status on “Remote Control” changing the Storage Charge/discharge Mode is almost immediately reflected in the inverter, functions as expected and is visible in the Mysolaredge app.

**Comment by biemond** [comment]
indeed, but at same point in the beginning the inverter wants to do test and after that it was allowing remote mode.

https://github.com/biemond/solaredge.modbus/blob/main/drivers/invertorwithbatt/device.ts#L459 plus I set it in remote mode 1st.

I think solaredge enabled it for you. mostly in some night batch after you try to do this.  Don't expect your installaltion guy is doing things proactive.

**Comment by CaptainVoni** [comment]
@kicksev  Could you tell me what firmware version you have on your inverter? Do you have the SetApp available?

**Comment by kicksev** [comment]
I don't have Setapp, but use the inverters wifi to browse to 172.16.0.1.
Firmware versions of my inverter:
CPU  4.17.136
DSP1  1.20.1551
Dsp2  2.20.1104

**Comment by CaptainVoni** [comment]
I got 4.17.136/1.20.1532/2.20.1103.

Can you check what you see under Commissioning, Power Control (the lightning icon), Energy Manager, Storage Control? I only have one entry there: AC Charge Limit. According to [this video](https://youtu.be/w2ap784QGO8?feature=shared&t=32), there should be another option "Grid Charge" that I simply don't have.

May I ask what country you are from?

**Comment by biemond** [comment]
menu is restricted . also for me. modbus allows me more. also if you want to see more you need an installation accounts.

**Comment by CaptainVoni** [comment]
Why don't you register such an account? I have two ;-)

**Comment by kicksev** [comment]
I am from Holland. My answer to the Commissioning question has to wait for a replacement of the communication board in the inverter. the system stopped communicating a few days ago, even the inverter wifi is not working.

**Comment by CaptainVoni** [comment]
@kicksev  Thanks for the info. I would be really interested to see what settings you have there. I'm pretty sure, that my inverter was able to charge from the grid up at the very beginning, but then it was disabled. I really wonder why and suspect some issue with (country-specific) regulation.

@biemond  You yourself might have given the answer to your own question ;-)

**Comment by biemond** [comment]
yep https://www.solar-bouwmarkt.nl/budget-omvormers-solaredge-monitoring-portal-in.html

**Comment by CaptainVoni** [comment]
I just got information from Solaredge directly: it is right now impossible to charge from the grid, it's not implemented yet. 😆 

I'm not sure they know how their products actually work...

**Comment by CaptainVoni** [comment]
> I am from Holland. My answer to the Commissioning question has to wait for a replacement of the communication board in the inverter. the system stopped communicating a few days ago, even the inverter wifi is not working.

Just checking: is your inverter still without communication? I'm looking for some "proof" that the Solaredge inverter actually can charge from the grid for my installer. So that he at least tries to find a way to enable it. :grin:

**Comment by kicksev** [comment]
@CaptainVoni My system is running again and I have looked at the Storage Control. Also in my system there is only 1 entry : AC charge limit which is set to no limit. To start charging and uncharging to AC I had to set the import and export limits to 5000 again,  (were reset to 0)

**Comment by CaptainVoni** [comment]
@kicksev Thanks for the hints. I tried pretty much everything, but I can't get it to charge. What exactly do you do to set the import and export limits? (Just to be safe we are talking about the same settings) Can you show me a flow where you set all the necessary setting to get it charging?

**Comment by kicksev** [comment]
@CaptainVoni  what country you are from? 
I have 2 48V batteries with a maximum charge/discharge load of 5000W

<img width="853" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/11637567-b469-4ec5-9493-b2cb80c5abd9">

To check the inverter and battery status: 
<img width="433" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/438822bb-5db0-4632-a460-57fae2b96781">

Inverter status:

0.	Undefined
1.	Off
2.	Sleeping
3.	Grid Monitoring
4.	Producing
5.	Producing (Throttled)
6.	Shutting Down'Fault
7.	Maintenance

Battery Status:

0.	Off
1.	Standby
2.	Initialising
3.	Charging
4.	Discharging
5.	Fault
6.	Idle
7.	Idle
8.	Idle
10.	Idle

**Comment by kicksev** [comment]
I will try to make a dashboard screen dump when my battery is loading from the grid at 5000W for your installer Too bad, at this moment the system is not listening to my commands, It does not react at all, may be I have played too much today.

**Comment by CaptainVoni** [comment]
That would be great, thanks.

As I wrote, right at the beginning I was able to charge from the grid. But that was when I installed and played around with this app and I think that playing-around triggered something to prevent future charging.

**Comment by kicksev** [comment]
@Biemond
On Oct 22 you showed a picture like this with charging 1000w and discharging 5000w, values my inverter is now also showing:
<img width="221" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/90e024ff-a99f-45aa-8c01-51ce095041a0">
You also showed a picture where charging and discharging settings were 5400W. How did you change these values? I tried setting charge power and discharge power with no success

**Comment by kicksev** [comment]
@CaptainVoni  I Think I start to understand the system a bit. My latest problem was caused by a minimum import limit set to 3000W. When set to zero, battery discharging started again. This limit seems to be the minimum power imported from the grid before the battery will discharge. As far as I can see are all modes of storage charge / discharge mode seem to work. This is a screendump when the battery is loaded with 4000W from AC according to the charge settings. 
<img width="763" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/84525b5c-c654-4b7b-8f45-850696b7a471">

**Comment by CaptainVoni** [comment]
@kicksev Thanks for all the info! I can't tell you how frustrating this is for me. No matter what I try, it doesn't charge from the grid. But now I have a screenshot from the Solaredge app I can pass along. Thanks again, I'll report back if I find any solution.

**Comment by kicksev** [comment]
@CaptainVoni Now testing the new TOU option in the Solaredge app. According to Solaredge: “_In Time of Use mode, we use your rate plan and smart forecasting and learning algorithm which optimizes your energy sources and battery usage to reduce your electric bills.
Based on your consumption habits and energy production, our algorithm predicts your consumption and energy generation potential for the following day, and dynamically charges or discharges the battery accordingly.
Time of Use mode automatically uses your battery when electricity rates are at their peak and solar is not available_.”
 In fact, at lowest hour price my battery is charged from the grid at full speed (5kW) but then it is discharged at random times. I am trying to find out what smart algorithms this behavior is based upon. But in your case, maybe the system will be forced to charge at 5kW.

**Comment by biemond** [comment]
oh, only was thinking they now support auto disabling solar export when hour price is negative 

is this a new option in your inverter menu?

**Comment by kicksev** [comment]
it in the Mysolaredge app. info: https://marketing.solaredge.com/tou-mode 
<img width="926" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/bde21801-7409-4f01-b2b5-7d395e3ca2fa">

**Comment by kicksev** [comment]
First week TOU. Charging starts at minimum hour price, discharging starts random? It does not look like the result of smart algorithms as promised.
[TOU test.pdf](https://github.com/biemond/solaredge.modbus/files/13845636/TOU.test.pdf)

**Comment by biemond** [comment]
interesting so they do this from their datacenter and gather all the data.  Did you see this menu option on your inverter? 
also like to see your change triggers, I guess they do the same trigger actions

**Comment by kicksev** [comment]
I have asked them where the algorithms run, but they told me to consult my installer who probably does not even know of the option. I saw the option on my mobile not on the inverter menu. What do you mean with change triggers?

**Comment by kicksev** [comment]
I have asked them where the algorithms run, but they told me to consult my installer who probably does not even know of the option. I do not understand your questions

**Comment by biemond** [comment]
oh, now I know this time of use.  I did this 2 years back . only NL as user you were not allowed to make these profiles, my installer did this for me and there were some profiles like day and night tariff

looks like they make it more advanced

**Comment by biemond** [comment]
[storedge_charge_discharge_profile_programming.pdf](https://github.com/biemond/solaredge.modbus/files/13845690/storedge_charge_discharge_profile_programming.pdf)

**Comment by biemond** [comment]
you need to click on the batt  and I see the same as the pdf but now from the app and I need to set all my peak moments. 
you can also do utility rate plan but I don't have this.so I got peak hours only

**Comment by kicksev** [comment]
In what country you are? Have you selected your tariff in the app?: 
<img width="1034" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/5fb4500f-2ac1-4b20-9dd3-b5eb0d59a3ab">

**Comment by biemond** [comment]
like you in NL but did not select dynamic, still do salderen.

**Comment by CaptainVoni** [comment]
@kicksev I tried the TOU option - didn't do anything. Just for fun I tried to charge from the grid:
![IMG_9169](https://github.com/biemond/solaredge.modbus/assets/1625329/0983a30a-0569-4c4c-8ea4-5b626f03b31c)

I didn't change anything. No update for the inverter. No idea - just works for now.

**Comment by kicksev** [comment]
The system is charging with 5kW. This should only take place at lowest rate hour. This screenshot should be taken at about 4 
 o'clock this morning, did you?  Please let me know if discharging will start at the peak hours (if you entered some). 

When I set the Storage charge/ discharge to Charge from PV+AC, it sometimes charges from PV+AC with 5kW and sometimes only with the available PV. Is there be a variable prohibiting/allowing charging from the grid besides the max discharge setting?

**Comment by kicksev** [comment]
in TOU mode last week my system charged from the grid at 5kW. This week it does charge from the grid at all, not by TOU mode nor when manually set to charge from PV+AC.  I wonder why. There is a register (Storage AC Charge Policy) enabling or disabling charging from the grid. do you know a way to read this register value or even set it?

Storage AC Charge Policy is used to enable charging for AC and the limit of yearly AC charge (if applicable).
0 – Disable
1 – Always allowed – needed for AC coupling operation. Allows unlimited charging from the AC. When used with Maximize
self-consumption, only excess power is used for charging (charging from the grid is not allowed).
2 – Fixed Energy Limit – allows AC charging with a fixed yearly (Jan 1 to Dec 31) limit (needed for meeting ITC regulation in
the US)
3 – Percent of Production - allows AC charging with a % of system production year to date limit (needed for meeting ITC
regulation in the US)

**Comment by CaptainVoni** [comment]
@kicksev This sounds like you found the reason why it sometimes charges and sometimes not! 🎉 Do you have a link to the document where you found that register mentioned? I only found it mentioned [here](https://github.com/evcc-io/evcc/discussions/6788) and it's stated that it is read-write, so we should be able to set it.

@biemond Can you add that register to the app? For the options 2 and 3 the following register (Storage AC Charge Limit) would be needed,  but I doubt that there is much use outside the US for those options.

**Comment by CaptainVoni** [comment]
@biemond The register "Storage Backup Reserved Setting" might also be worth considering. It seems to relate to the minimum SOC of the battery.

**Comment by biemond** [comment]
let me check the docs which register this is plus what is my value for this.

**Comment by biemond** [comment]
can you send me diagnostic reports, I already added this 

```
        "storage_control_mode": [0xe004, 1, 'UINT16', "Storage Control Mode"],
        "storage_accharge_policy": [0xe005, 1, 'UINT16', "Storage AC Charge Policy"],
        "storage_accharge_Limit": [0xe006, 2, 'FLOAT32', "Storage AC Charge Limit"],
```

**Comment by CaptainVoni** [comment]
07d74748-6c83-4ab2-b214-10d109151f0f

**Comment by biemond** [comment]
```
storage_control_mode 4 xxx Storage Control Mode
storage_accharge_policy 2 xxx Storage AC Charge Policy
storage_accharge_Limit 0 xxx Storage AC Charge Limit
```

**Comment by CaptainVoni** [comment]
With what version did you add that?

**Comment by biemond** [comment]
what do you mean ? I just already read this 

2 – Fixed Energy Limit – allows AC charging with a fixed yearly (Jan 1 to Dec 31) limit (needed for meeting ITC regulation in
the US)

**Comment by biemond** [comment]
maybe in the new year you have some new quota.

**Comment by biemond** [comment]
it was always there , never showed this on the device.

**Comment by CaptainVoni** [comment]
> maybe in the new year you have some new quota.

That was also my first thought. But at least we might have an idea right now what’s going on.

**Comment by biemond** [comment]
will make a device action so you can change it.

**Comment by biemond** [comment]
nop , I never touch or set this. just reading. there is no code which can do this on the app.

**Comment by biemond** [comment]
I got these 
```
storage_control_mode 1 xxx Storage Control Mode
storage_accharge_policy 1 xxx Storage AC Charge Policy
storage_accharge_Limit 0 xxx Storage AC Charge Limit
remote_control_command_mode -1 xxx Remote Control Command Mode
remote_control_charge_limit 5000 xxx Remote Control Charge Limit
remote_control_command_discharge_limit 5000 xxx Remote Control Command Discharge Limit
remote_control_command_timeout 3600 xxx Remote Control Command Timeout
remote_control_default_command_mode 0 xxx Storage Charge/Discharge Default Mode
```

**Comment by biemond** [comment]
I think a hour is the default for reset. 

I do more 

```
      if (type == 'storagedefaultmode') {
        const storagedefaultmodeRes = await client.writeSingleRegister(0xe004, Number(4));
        console.log('controlmodewrite', storagedefaultmodeRes)

        // set timeout to 6 hours, done in seconds
        const remoteTimeout = await client.writeMultipleRegisters(0xe00b, [ Number(21600), 0 ]);
        console.log('remote_control_command_timeout', remoteTimeout);
```

**Comment by kicksev** [comment]
@CaptainVoni @biemond I may be too late with this link to the document you asked for: https://www.photovoltaikforum.com/core/attachment/88445-power-control-open-protocol-for-solaredge-inverters-pdf/ 
I see Biemond wrote some code to set some registers. Although I have programmed a lot of my life, I'm not familiar with this coding, so could you tell me what is done and if this can help me to charge from AC?

**Comment by CaptainVoni** [comment]
@kicksev If I’m not mistaken, @biemond will update the app in the next time, so that it will be possible to change the policy setting.

Patience, young padawan! 😉

**Comment by biemond** [comment]
I think we don't care  about Storage AC Charge Limit  if we can set this Storage AC Charge Policy   


Storage AC Charge Limit is used to set the AC charge limit according to the policy set in the previous register. Either fixed in kWh or percentage is set (e.g. 100KWh or 70%). Relevant only for Storage AC Charge Policy = 2 or 3.

**Comment by biemond** [comment]
Can you try the menu action, it works for me https://homey.app/a/solaredge.modbus/test/
<img width="486" alt="Screen Shot 2024-01-14 at 17 30 27" src="https://github.com/biemond/solaredge.modbus/assets/1085004/b5f8f73d-1480-42d8-904e-247b68decba6">
<img width="471" alt="Screen Shot 2024-01-14 at 17 30 34" src="https://github.com/biemond/solaredge.modbus/assets/1085004/213b76c4-2d4f-4866-beee-887e4e8bda2b">

**Comment by CaptainVoni** [comment]
@biemond IT WORKS! 🎉 Thank you so much for solving this!
@kicksev Also a big thank you to you for identifying the problem!

**Comment by biemond** [comment]
cool , in a few days when I don't hear any issues , I will publish it to release..

how did you like my German?  bit rusty on that but google helped me a bit.

**Comment by biemond** [comment]
By the way is device action enough or do you want a flow action?

**Comment by CaptainVoni** [comment]
> cool , in a few days when I don't hear any issues , I will publish it to release..

In the meantime I added a flow the charge every day at the cheapest hour just for testing purposes. I will let this activated for a week or so to see if any problems arise.
 
> how did you like my German? bit rusty on that but google helped me a bit.

Pretty fine. The labels for the options are all perfect, for the policy I would use "Speicher-AC-Laderegelung" instead. If you need German translations for any project, just let me know - I'm happy to help.

> By the way is device action enough or do you want a flow action?

I'm not sure. Do we know if this setting will revert to some default after some time? Or after a restart of the inverter? Or an update? If a flow action would be available, I would set the policy every time I activate grid charging - just to be on the safe side.

@kicksev Just out of curiosity: what policy option was selected for you? At my inverter it was set to "Fixed Energy Limit".

@biemond Did I already say thank you? Thank you!

**Comment by kicksev** [comment]
@biemond Thx for your support, I'm very happy with this option.
I set mode to “Always allowed” and “charging from PV + AC” immediately started at 5kW, then I set “Maximize export” and discharging at 5kW started after a few seconds. Back to “Max Self consumption” was also OK. I agree with CaptainVoni according the need for an action.
@CaptainVoni,  As I updated the app in homey the Storage AC Charge Policy was already set to Always allowed. 
the TOU option of Solaredge is useless, If from now on charging from AC keeps working, I will optimize the system through Homey. I made also some flows. I fully charge at cheapest hour at night, set Disabled when full and start Max Self Consumption at most expensive hour. If there is not enough PV power to power the house and car, this will do the job.

**Comment by CaptainVoni** [comment]
@biemond Just to report back: battery is charging without any problem every day now. I couldn't spot any other problems either, so for me it's fine 👍

**Comment by kicksev** [comment]
@biemond @CaptainVoni  If the status is charge PV+AC (by Homey or TOU in app) then at 5:00, the status is suddenly changed to Max Self Consumption although the battery is not fully loaded yet. Any other time of the day this does not happen.  Any idea where this comes from?
<img width="1138" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/96e23cbf-3112-403b-9fff-429aa1fdef6d">

**Comment by CaptainVoni** [comment]
@kicksev What happens if you reset the status to PV+AC at 5:01? Does it continue to charge? Do you set the status always at 4:00 and it resets at 5:00? What happens if you set the status at 3:30? Does it reset at 5:00 or 4:30? It might have something to do with the timeout of remote commands.

**Comment by biemond** [comment]
it can this timeout after you do a custom action, I think my timeout is 6 hours plus never exactly on the hour.  But I think Solaredge is doing this forced because they think it is wise and mostly more expensive to charge from grid. maybe some cron on their side

**Comment by kicksev** [comment]
@biemond, @CaptainVoni Homey sets the mode to charge PV+AC at the time when power price is at lowest and releases the battery power when price is at highest. At this moment it is only saving about 1 euro per day but I suppose that the in day price difference will grow in the future and I will be prepared.
Homey sends me an email every time one of the 4 statuses of the PV system changes. I just had a good look at them, and I saw that now and then the inverter goes into maintenance mode and at the same time the charge/discharge mode is set to MSC. This happens almost every day around 5:00 AM, 19:00 PM and sometimes in between. Solaredge does not reset the statuses to the previous value. It is probably a Solaredge bug and I have to find a workaround. 
In the screenshots you see that Thursday charging was started at 1:01 am and complete at 2:55 am (status set at Off by Homey) and after some I don't know why status changes at 5:00 am maintenance started. On the screenshot of today you can see that charging started at 4:01 am and maintenance at 5:00 am and that charging stopped as mode was set to MSC.

Thursday
![image](https://github.com/biemond/solaredge.modbus/assets/10222271/d2d2213a-f34b-4c76-b241-0f38efb7c48f)

Friday
<img width="881" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/4e46be3d-b670-4fb8-a521-2afa5dec34f7">

**Comment by biemond** [comment]
interesting -> inverter goes into maintenance mode
let me check on my side if I have also these maintenance modes.

**Comment by CaptainVoni** [comment]
I added a flow that logs any status changes of the inverter to the timeline. I don’t see any maintenance on my side, so this might be specific for your inverter.

Regarding your cost savings: I know, no one wants to hear this - me included - but did you consider the wear and tear of the battery when calculating your savings? I did that based on the guaranteed amount of energy that can be taken from the battery during its whole life and ended up with about 0.20€/kWh. That’s just the cost for the actual process of charging and discharging the battery with one kWh and without the cost of that one kWh you have to charge the battery with before you get it out again from the battery. I ended up with using just 0.10€/kWh in my battery optimisation flows, otherwise the battery won’t be used at all. My current costs for electricity out of the battery are 0.182€/kWh, but only if charged only from PV (0.10€ for wear and tear and 0.082€ for electricity from PV).

**Comment by biemond** [comment]
same for me , just producing, offline , no maintenance.

**Comment by kicksev** [comment]
@CaptainVoni  I did not make that calculation, but I fully agree that batteries do not render. How many times the batteries can be charged and discharged and what is the charging/discharging efficiency in your calculation? 
I bought my batteries primarily for backup as I expect blackouts in the future due to a failing energy transition. Besides that, I am curious to find out how much extra PV power my batteries can save from sending to the grid. What is the main reason for you to buy batteries knowing they will not make you rich?

**Comment by biemond** [comment]
yep and you loose 10% in all the conversions but for me it is a hobby . also like to have this bi-directional charger plus this backup system of solaredge.  maybe after 2025 will do dynamic tariff.

**Comment by CaptainVoni** [comment]
@kicksev  For me it’s the same as for @biemond : mainly a hobby. I want to see what is possible, knowing that overall it will cost me a bit more instead of actually saving money.

I based my calculation on (i) the guaranteed amount of electricity I can get out of the battery according to its data sheet and (ii) the price I paid for the battery. That gave me around 0.20€/kWh. As I hope that I will get a bit more out of it than the guaranteed amount and that battery prices will have fallen when I have to replace it, I use just 0.10€/kWh in my optimisation strategy. This of course means that the battery will only be charged from the grid if the difference between the highest and lowest price is more than 0.10€, which is very rare these days, e.g. today is the difference only 0.03€.

**Comment by biemond** [comment]
after a week , I never had maintenance , only sleeing , producing or producing throttled

**Comment by kicksev** [comment]
@biemond , @CaptainVoni  My system data and version numbers are 
<img width="199" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/09dbaabb-7dbd-41c7-846e-a906e39044e3">
<img width="143" alt="image" src="https://github.com/biemond/solaredge.modbus/assets/10222271/a4026c3a-cd83-4e4b-9047-19ba7eb84b01">

What system do you have?

**Comment by biemond** [comment]
Model | SE4000H-RWS00BNO4  
Communication Board (CPU) Version | 4.17.218
DSP2 Version | 2.0.1305
DSP1 Version | 1.0.1616

SolarEdge Optimizer P370


LG Chem RESU 10kWh

**Comment by CaptainVoni** [comment]
See below my previous and updated versions.
![image](https://github.com/biemond/solaredge.modbus/assets/1625329/9b35e04b-610b-4957-89ed-20c6c784c97b)

@biemond FYI: it seems like that the control mode “Remote control” was replaced by “Remote Control + TOU”. Right now I don’t see any issues.
